File: RotateALinkedList/RotateALinkedList.cpp
/*
https://www.geeksforgeeks.org/problems/rotate-a-linked-list/1
*/

//{ Driver Code Starts
// Initial Template for C++
#include <bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    struct Node* next;

    Node(int x) {
        data = x;
        next = NULL;
    }
};

void printList(Node* node) {
    while (node != NULL) {
        cout << node->data << " ";
        node = node->next;
    }
    cout << "\n";
}


// } Driver Code Ends
/*

struct Node {
    int data;
    struct Node *next;
    Node(int x) {
        data = x;
        next = NULL;
    }
};

*/

class Solution {
  public:
    // Function to rotate a linked list.
    Node* rotate(Node* head, int k) {
        vector<int>others;
        Node *toad=head;
        Node *toadTail=toad;
        while(true){
            others.push_back(toadTail->data);
            if(toadTail->next==nullptr&&k!=0) return toad;
            toadTail=toadTail->next;
            k--;
            if(k==0) break;
        }
        Node *tmphead=toadTail;
        Node *tailtmphead=tmphead;
        while(true){
            if(tailtmphead->next==nullptr) break;
            tailtmphead=tailtmphead->next;
        }
        for(auto &vals: others){
            Node *nwNode=new Node(vals);
            tailtmphead->next=nwNode;
            tailtmphead=tailtmphead->next;
        }
        return tmphead;
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    cin.ignore();

    while (t--) {
        vector<int> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int number;

        // Read numbers from the input line
        while (ss >> number) {
            arr.push_back(number);
        }

        Node* head = nullptr;

        // Check if the array is empty
        if (!arr.empty()) {
            head = new Node(arr[0]);
            Node* tail = head;
            for (size_t i = 1; i < arr.size(); ++i) {
                tail->next = new Node(arr[i]);
                tail = tail->next;
            }
        }
        int k;
        cin >> k;
        cin.ignore();

        Solution ob;
        head = ob.rotate(head, k);
        printList(head);
    }
    return 1;
}

// } Driver Code Ends

File: RotateALinkedList/RotateALinkedList.java
/*
https://www.geeksforgeeks.org/problems/rotate-a-linked-list/1
*/

//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class Node {
    int data;
    Node next;

    Node(int x) {
        data = x;
        next = null;
    }
}


// } Driver Code Ends
/* node of linked list:

class Node{
    int data;
    Node next;
    Node(int d){
        data=d;
        next=null;
    }
}

*/

class Solution {
    // Function to rotate a linked list.
    public Node rotate(Node head, int k) {
        ArrayList<Integer>listNodesAfterLeftShift=new ArrayList<>();
        Node temporaryHead=head;
        Node tailForTemporaryHead=temporaryHead;
        while(true){
            listNodesAfterLeftShift.add(tailForTemporaryHead.data);
            if(tailForTemporaryHead.next==null&&k!=0) return temporaryHead;
            tailForTemporaryHead=tailForTemporaryHead.next;
            k--;
            if(k==0) break;
        }
        Node newHeadForTemporaryTail=tailForTemporaryHead;
        Node tailForNewHead=newHeadForTemporaryTail;
        while(true){
            if(tailForNewHead.next==null) break;
            tailForNewHead=tailForNewHead.next;
        }
        for(int i=0; i<listNodesAfterLeftShift.size(); ++i){
            Node newNode=new Node(listNodesAfterLeftShift.get(i));
            tailForNewHead.next=newNode;
            tailForNewHead=tailForNewHead.next;
        }
        return newHeadForTemporaryTail;
    }
}


//{ Driver Code Starts.

public class GFG {
    static void printList(Node node) {
        while (node != null) {
            System.out.print(node.data + " ");
            node = node.next;
        }
        System.out.println();
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine().trim());

        while (t-- > 0) {
            List<Integer> arr = new ArrayList<>();
            String input = br.readLine().trim();
            if (!input.isEmpty()) {
                String[] numbers = input.split("\\s+");
                for (String num : numbers) {
                    if (!num.isEmpty()) {
                        arr.add(Integer.parseInt(num));
                    }
                }
            }

            Node head = null;
            if (!arr.isEmpty()) {
                head = new Node(arr.get(0));
                Node tail = head;
                for (int i = 1; i < arr.size(); ++i) {
                    tail.next = new Node(arr.get(i));
                    tail = tail.next;
                }
            }
            int k = Integer.parseInt(br.readLine().trim());
            Solution ob = new Solution();
            head = ob.rotate(head, k);
            printList(head);
        }
    }
}

// } Driver Code Ends